usr.bin/passwd/passwd.1
.\" Copyright (c) 1990 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	from: @(#)passwd.1	6.11 (Berkeley) 7/24/91
.\"	$Id: passwd.1,v 1.1 1995/10/18 08:45:54 deraadt Exp $
.\"
.Dd July 24, 1991
.Dt PASSWD 1
.Os BSD 4
.Sh NAME
.Nm passwd
.Nd modify a user's password
.Sh SYNOPSIS
.Nm passwd
.Op Fl l
.Op Fl k
.Op Fl y
.Op Ar user
.Nm yppasswd
.Op Ar user
.Sh DESCRIPTION
.Nm Passwd
changes the user's local, Kerberos, or YP password.  First, the user is prompted
for their current password.
If the current password is correctly typed, a new password is
requested.
The new password must be entered twice to avoid typing errors.
.Pp
The new password should be at least six characters long and not
purely alphabetic.
Its total length must be less than
.Dv _PASSWORD_LEN
(currently 128 characters).
Numbers, upper case letters and meta characters
are encouraged.
.Bl -tag -width flag
.It Fl l
This option causes the password to be updated only in the local
password file.  When changing only the local password,
.Xr pwd_mkdb  8
is used to update the password databases.
.It Fl k
This option forces the change to affect the Kerberos database, even
if the user has a password in the local database.
Once the password has been verified,
.Nm passwd
communicates the new password information to
the Kerberos authenticating host.
.It Fl y
This forces the YP password database entry to be changed, even if
the user has an entry in the local database.  The
.Xr rpc.yppasswdd 8
daemon should be running on the YP master server.
.Nm yppasswd
is the equivalent of
.Nm passwd
with the
.Fl y
flag.
.El
.Pp
This is the behavior if no flags are specified:
if Kerberos is active then
.Nm passwd
will talk to the Kerberos server (even if the user has an entry
in the local database.)
If the password is not in the local password database, then
an attempt is made to use the YP database.
.Pp
To change another user's Kerberos password, one must first
run
.Xr kinit 1
followed by
.Xr passwd 1 .
The super-user is not required to provide a user's current password
if only the local password is modified.
.Sh FILES
.Bl -tag -width /etc/master.passwd -compact
.It Pa /etc/master.passwd
The user database
.It Pa /etc/passwd 
A Version 7 format password file
.It Pa /etc/passwd.XXXXXX
Temporary copy of the password file
.El
.Sh SEE ALSO
.Xr chpass 1 ,
.Xr kerberos 1 ,
.Xr kinit 1 ,
.Xr login 1 ,
.Xr passwd 5 ,
.Xr kpasswdd 8 ,
.Xr pwd_mkdb 8 ,
.Xr vipw 8
.Rs
.%A Robert Morris
.%A Ken Thompson
.%T "UNIX password security"
.Re
.Sh HISTORY
A
.Nm passwd
command appeared in
.At v6 .


usr.bin/passwd/passwd.c
/*
 * Copyright (c) 1988 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1988 The Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */

#ifndef lint
/*static char sccsid[] = "from: @(#)passwd.c	5.5 (Berkeley) 7/6/91";*/
static char rcsid[] = "$Id: passwd.c,v 1.1 1995/10/18 08:45:54 deraadt Exp $";
#endif /* not lint */

#include <stdio.h>
#include <string.h>
#include <unistd.h>

/*
 * Note on configuration:
 *      Generally one would not use both Kerberos and YP
 *      to maintain passwords.
 *
 */

int use_kerberos;
int use_yp;

#ifdef YP
int force_yp;
#endif

main(argc, argv)
	int argc;
	char **argv;
{
	extern int optind;
	register int ch;
	char *username;
	int status = 0;
	char *basename;

#if defined(KERBEROS) || defined(KERBEROS5)
	use_kerberos = 1;
#endif
#ifdef	YP
	use_yp = _yp_check(NULL);
#endif

	basename = strrchr(argv[0], '/');
	if (basename == NULL)
		basename = argv[0];
	if (strcmp(basename, "yppasswd") == 0) {
#ifdef YP
		if (!use_yp) {
			fprintf(stderr, "yppasswd: YP not in use.\n");
			exit (1);
		}
		use_kerberos = 0;
		use_yp = 1;
		force_yp = 1;
#else
		fprintf(stderr, "yppasswd: YP not compiled in\n");
		exit(1);
#endif
	}

	
	while ((ch = getopt(argc, argv, "lky")) != EOF)
		switch (ch) {
		case 'l':		/* change local password file */
			use_kerberos = 0;
			use_yp = 0;
			break;
		case 'k':		/* change Kerberos password */
#if defined(KERBEROS) || defined(KERBEROS5)
			use_kerberos = 1;
			use_yp = 0;
			break;
#else
			fprintf(stderr, "passwd: Kerberos not compiled in\n");
			exit(1);
#endif
		case 'y':		/* change YP password */
#ifdef	YP
			if (!use_yp) {
				fprintf(stderr, "passwd: YP not in use.\n");
				exit(1);
			}
			use_kerberos = 0;
			use_yp = 1;
			force_yp = 1;
			break;
#else
			fprintf(stderr, "passwd: YP not compiled in\n");
			exit(1);
#endif
		default:
			usage();
			exit(1);
		}

	argc -= optind;
	argv += optind;

	username = getlogin();
	if (username == NULL) {
		fprintf(stderr, "passwd: who are you ??\n");
		exit(1);
	}
	
	switch(argc) {
	case 0:
		break;
	case 1:
#if defined(KERBEROS) || defined(KERBEROS5)
		if (use_kerberos && strcmp(argv[0], username)) {
			(void)fprintf(stderr, "passwd: %s\n\t%s\n%s\n",
"to change another user's Kerberos password, do",
"\"kinit <user>; passwd; kdestroy\";",
"to change a user's local passwd, use \"passwd -l <user>\"");
			exit(1);
		}
#endif
		username = argv[0];
		break;
	default:
		usage();
		exit(1);
	}

#if defined(KERBEROS) || defined(KERBEROS5)
	if (use_kerberos)
		exit(krb_passwd());
#endif
#ifdef	YP
	if (force_yp || ((status = local_passwd(username)) && use_yp))
		exit(yp_passwd(username));
	exit(status);
#endif
	exit(local_passwd(username));
}

usage()
{
	fprintf(stderr, "usage: passwd [-l] [-k] [-y] user\n");
}
